File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseMatchPlayerRepository.java
package com.yahyeet.boardbook.model.firebase.repository;

import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QuerySnapshot;
import com.yahyeet.boardbook.model.entity.MatchPlayer;
import com.yahyeet.boardbook.model.repository.IMatchPlayerRepository;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionException;
import java.util.stream.Collectors;

/**
 * Firebase Firestore implementation of a match player repository
 */
public class FirebaseMatchPlayerRepository extends AbstractFirebaseRepository<MatchPlayer> implements IMatchPlayerRepository {
	private Map<String, List<String>> matchMatchPlayerIdsCache = new HashMap<>();
	private Map<String, List<String>> userMatchPlayerIdsCache = new HashMap<>();

	public FirebaseMatchPlayerRepository(FirebaseFirestore firestore) {
		super(firestore);
	}

	@Override
	public AbstractFirebaseEntity<MatchPlayer> fromModelEntityToFirebaseEntity(MatchPlayer entity) {
		return FirebaseMatchPlayer.fromModelType(entity);
	}

	@Override
	public AbstractFirebaseEntity<MatchPlayer> fromDocumentToFirebaseEntity(DocumentSnapshot document) {
		return FirebaseMatchPlayer.fromDocument(document);
	}

	@Override
	public String getCollectionName() {
		return "match_players";
	}

	@Override
	public CompletableFuture<List<MatchPlayer>> findMatchPlayersByMatchId(String id) {
		List<String> cachedMatchPlayerIds = matchMatchPlayerIdsCache.get(id);
		if (cachedMatchPlayerIds != null) {
			List<CompletableFuture<MatchPlayer>> futureMatchPlayers =
				cachedMatchPlayerIds
					.stream()
					.map(this::find)
					.collect(Collectors.toList());

			return CompletableFuture.allOf(futureMatchPlayers.toArray(new CompletableFuture[0]))
				.thenApply(future -> futureMatchPlayers.stream().map(CompletableFuture::join).collect(Collectors.toList()));
		}

		return CompletableFuture.supplyAsync(() -> {
			Task<QuerySnapshot> task =
				getFirestore()
					.collection(getFullCollectionName())
					.whereEqualTo("matchId", id)
					.get();

			try {
				QuerySnapshot querySnapshot = Tasks.await(task);

				return querySnapshot
					.getDocuments()
					.stream()
					.map(this::fromDocumentToFirebaseEntity)
					.collect(Collectors.toList());
			} catch (Exception e) {
				throw new CompletionException(e);
			}
		}).thenApply(firebaseMatchPlayers -> {
			firebaseMatchPlayers.forEach(firebaseMatchPlayer -> {
				getCache().put(firebaseMatchPlayer.getId(), firebaseMatchPlayer);
			});

			matchMatchPlayerIdsCache.put(id, firebaseMatchPlayers.stream().map(AbstractFirebaseEntity::getId).collect(Collectors.toList()));

			return firebaseMatchPlayers
				.stream()
				.map(AbstractFirebaseEntity::toModelType)
				.collect(Collectors.toList());
		});
	}

	@Override
	public CompletableFuture<List<MatchPlayer>> findMatchPlayersByUserId(String id) {
		List<String> cachedMatchPlayerIds = userMatchPlayerIdsCache.get(id);
		if (cachedMatchPlayerIds != null) {
			List<CompletableFuture<MatchPlayer>> futureMatchPlayers =
				cachedMatchPlayerIds
					.stream()
					.map(this::find)
					.collect(Collectors.toList());

			return CompletableFuture.allOf(futureMatchPlayers.toArray(new CompletableFuture[0]))
				.thenApply(future -> futureMatchPlayers.stream().map(CompletableFuture::join).collect(Collectors.toList()));
		}

		return CompletableFuture.supplyAsync(() -> {
			Task<QuerySnapshot> task =
				getFirestore()
					.collection(getFullCollectionName())
					.whereEqualTo("playerId", id)
					.get();

			try {
				QuerySnapshot querySnapshot = Tasks.await(task);

				return querySnapshot
					.getDocuments()
					.stream()
					.map(this::fromDocumentToFirebaseEntity)
					.collect(Collectors.toList());
			} catch (Exception e) {
				throw new CompletionException(e);
			}
		}).thenApply(firebaseMatchPlayers -> {
			firebaseMatchPlayers.forEach(firebaseMatchPlayer -> {
				getCache().put(firebaseMatchPlayer.getId(), firebaseMatchPlayer);
			});

			userMatchPlayerIdsCache.put(id, firebaseMatchPlayers.stream().map(AbstractFirebaseEntity::getId).collect(Collectors.toList()));

			return firebaseMatchPlayers
				.stream()
				.map(AbstractFirebaseEntity::toModelType)
				.collect(Collectors.toList());
		});
	}

	@Override
	public CompletableFuture<Void> afterSave(MatchPlayer entity, AbstractFirebaseEntity<MatchPlayer> savedEntity) {
		return CompletableFuture.supplyAsync(() -> {
			List<String> cachedMatchMatchPlayerIds = matchMatchPlayerIdsCache.get(entity.getId());

			if (cachedMatchMatchPlayerIds != null && cachedMatchMatchPlayerIds.stream().noneMatch(matchPlayerId -> matchPlayerId.equals(entity.getId()))) {
				cachedMatchMatchPlayerIds.add(entity.getId());
			}

			List<String> cachedUserMatchPlayerIds = userMatchPlayerIdsCache.get(entity.getId());

			if (cachedUserMatchPlayerIds != null && cachedUserMatchPlayerIds.stream().noneMatch(matchPlayerId -> matchPlayerId.equals(entity.getId()))) {
				cachedUserMatchPlayerIds.add(entity.getId());
			}

			return null;
		});
	}

	@Override
	public CompletableFuture<Void> afterDelete(String matchPlayerId) {
		return CompletableFuture.supplyAsync(() -> {
			for (Map.Entry<String, List<String>> pair : matchMatchPlayerIdsCache.entrySet()) {
				if (pair.getValue().stream().anyMatch(cachedMatchPlayerId -> cachedMatchPlayerId.equals(matchPlayerId))) {
					matchMatchPlayerIdsCache.put(pair.getKey(), pair.getValue().stream().filter(cachedMatchPlayerId -> !cachedMatchPlayerId.equals(matchPlayerId)).collect(Collectors.toList()));
					break;
				}
			}

			for (Map.Entry<String, List<String>> pair : userMatchPlayerIdsCache.entrySet()) {
				if (pair.getValue().stream().anyMatch(cachedMatchPlayerId -> cachedMatchPlayerId.equals(matchPlayerId))) {
					userMatchPlayerIdsCache.put(pair.getKey(), pair.getValue().stream().filter(cachedMatchPlayerId -> !cachedMatchPlayerId.equals(matchPlayerId)).collect(Collectors.toList()));
					break;
				}
			}

			return null;
		});
	}
}

File: application/app/src/main/java/com/yahyeet/boardbook/model/firebase/repository/FirebaseMatchPlayerRepository.kt
package com.yahyeet.boardbook.model.firebase.repository

import com.google.android.gms.tasks.Task
import com.google.android.gms.tasks.Tasks
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.QuerySnapshot
import com.yahyeet.boardbook.model.entity.MatchPlayer
import com.yahyeet.boardbook.model.repository.IMatchPlayerRepository
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionException
import java.util.stream.Collectors

class FirebaseMatchPlayerRepository(firestore: FirebaseFirestore) : AbstractFirebaseRepository<MatchPlayer>(firestore), IMatchPlayerRepository {

    override val collectionName: String
        get() = "match_players"

    override fun fromModelEntityToFirebaseEntity(entity: MatchPlayer): AbstractFirebaseEntity<MatchPlayer> {
        return FirebaseMatchPlayer.fromModelType(entity)
    }

    override fun fromDocumentToFirebaseEntity(document: DocumentSnapshot): AbstractFirebaseEntity<MatchPlayer> {
        return FirebaseMatchPlayer.fromDocument(document)
    }

    override fun findMatchPlayersByMatchId(id: String): CompletableFuture<List<MatchPlayer>> {
        return CompletableFuture.supplyAsync<List<AbstractFirebaseEntity<MatchPlayer>>> {
            val task = firestore
                    .collection(collectionName)
                    .whereEqualTo("matchId", id)
                    .get()

            try {
                val querySnapshot = Tasks.await(task)

                return@CompletableFuture.supplyAsync querySnapshot
                        .getDocuments()
                        .stream()
                        .map(Function<DocumentSnapshot, AbstractFirebaseEntity<MatchPlayer>> { this.fromDocumentToFirebaseEntity(it) })
                        .collect(Collectors.toList<AbstractFirebaseEntity<MatchPlayer>>())
            } catch (e: Exception) {
                throw CompletionException(e)
            }
        }.thenApply { firebaseMatchPlayers ->
            firebaseMatchPlayers.forEach { firebaseMatchPlayer -> cache[firebaseMatchPlayer.id!!] = firebaseMatchPlayer }

            firebaseMatchPlayers
                    .stream()
                    .map<MatchPlayer>(Function<AbstractFirebaseEntity<MatchPlayer>, MatchPlayer> { it.toModelType() })
                    .collect<List<MatchPlayer>, Any>(Collectors.toList())
        }
    }

    override fun findMatchPlayersByUserId(id: String): CompletableFuture<List<MatchPlayer>> {
        return CompletableFuture.supplyAsync<List<AbstractFirebaseEntity<MatchPlayer>>> {
            val task = firestore
                    .collection(collectionName)
                    .whereEqualTo("playerId", id)
                    .get()

            try {
                val querySnapshot = Tasks.await(task)

                return@CompletableFuture.supplyAsync querySnapshot
                        .getDocuments()
                        .stream()
                        .map(Function<DocumentSnapshot, AbstractFirebaseEntity<MatchPlayer>> { this.fromDocumentToFirebaseEntity(it) })
                        .collect(Collectors.toList<AbstractFirebaseEntity<MatchPlayer>>())
            } catch (e: Exception) {
                throw CompletionException(e)
            }
        }.thenApply { firebaseMatchPlayers ->
            firebaseMatchPlayers.forEach { firebaseMatchPlayer -> cache[firebaseMatchPlayer.id!!] = firebaseMatchPlayer }

            firebaseMatchPlayers
                    .stream()
                    .map<MatchPlayer>(Function<AbstractFirebaseEntity<MatchPlayer>, MatchPlayer> { it.toModelType() })
                    .collect<List<MatchPlayer>, Any>(Collectors.toList())
        }
    }
}
